change notifications and change tracking patterns (#302)
Microsoft/api-guidelines: graph/patterns/change-tracking.md

*The change tracking pattern provides the ability to keep API consumers in sync with changes in Microsoft Graph without having to continuously poll the API.*

## Problem
---------

API consumers require an efficient way to keep data in sync with Microsoft Graph, and the API design should not allow for continuous polling as it'd be costly and wouldn't guarantee data integrity.

## Solution
--------

API designers can enable the change tracking (delta) capability on entity collections by declaring a delta function for API consumers to use to track changes in that collection.

This new endpoint can be used to sync API consumers. This is achieved through returning a delta link with a watermark. Once the API consumer needs to refresh the data it uses the last provided delta link to catch up on new changes since their last request. Delta guarantees integrity of data through the watermark, regardless of service partitions and other obscure aspects for clients.

## Issues and Considerations
-------------------------

Implementer MUST implement a watermark storage system in case of active watermarks (cursor in data store, partition affinity, sync state in data store...).

> ## When to Use this Pattern

Author: ```suggestion
> Note: the delta function also provides support for $filter and $select to allow the API consumer to narrow down the number of entities and properties retrieved as well as the number of changes that are tracked. Additionally the delta function can also support $top to allow the API consumer to sync smaller sets of changes as well as $expand to allow the API consumer to sync related data. Expand across workloads is not supported today however.

## When to Use this Pattern
```

## Solution
--------

API designers can enable the change tracking (delta) capability on entity collections by declaring a delta function for API consumers to use to track changes in that collection.

This new endpoint can be used to sync API consumers. This is achieved through returning a delta link with a watermark. Once the API consumer needs to refresh the data it uses the last provided delta link to catch up on new changes since their last request. Delta guarantees integrity of data through the watermark, regardless of service partitions and other obscure aspects for clients.

> Note: although this capability is similar to the [OData $delta feed](https://docs.oasis-open.org/odata/odata-json-format/v4.0/errata02/os/odata-json-format-v4.0-errata02-os-complete.html#_Toc403940644) capability, it is a different construct. Microsoft Graph APIs MUST provide change tracking through the delta function and MUST NOT implement the OData $delta feed when providing change tracking capabilities to ensure the uniformity of the API experience.

## Issues and Considerations
-------------------------

> Implementer MUST implement a watermark storage system in case of active watermarks. Passive watermarks are watermarks that can be retrieved from the context (e.g. timestamp), active watermarks represent information required to track the sync state which cannot be retrieved from the context (e.g. cursor from data store, partition affinity marker, partition id, generated unique sync identifier...)

Reviewer: You know I'm not sure we shoudl; use MUST language when we're talking about implementation choices.  These guidelines should confine themselves on the whole to what behavior is observable from the outside caller.

If the cursor is fully encoded and internalized in the delta next token then no internal storage is required.

Author: yeah that's the point, because the cursor is already stored, no need to store it twice and the workload would satisfy the requirement.

> 

Author: ```suggestion

Implementer MUST implement soft deletion for entities in the backend storage system. The soft deletion will provide useful information to the client to appropriately reflect deletions.

When an entity is soft deleted, the delta function MUST return the id of the deleted entity as well as a `@removed` annotation with the `reason` field.
- The reason MUST be set to `changed` if the entity can be restored. `"@removed": {"reason": "changed"}`. 
- The reason MUST be set to `deleted` if the entity cannot be restored. `"@removed": {"reason": "deleted"}`.

When a link to an entity is deleted, or when the linked entity is deleted, or when a link to an entity is added, implementer MUST return a `property@delta` annotation. e.g. considering the entity Group has a navigation property named members of type Collection(user):

- When a user is added to the group `"members@delta": ["id of the added user"]`
- When a user is removed from the group, or the target user is deleted `"members@delta": ["id of the added user"]`

```

> Note: the delta function also provides support for $filter and $select to allow the API consumer to narrow down the number of entities and properties retrieved as well as the number of changes that are tracked. Additionally the delta function can also support $top to allow the API consumer to sync smaller sets of changes as well as $expand to allow the API consumer to sync related data. Expand across workloads is not supported today however.

> 

Reviewer: Should document the  "manager@delta": [] for nav properties part of the pattern here.

Author: added above

## When to Use this Pattern
------------------------

Before using the change tracking pattern in your API definition, make sure your scenario fits the following criteria:

- API consumers want to sync the data.

> - API consumers don't want to be immediately notified of changes. (see change notifications pattern for this scenario).

Reviewer: This isn't quite correct.   Cos when people are doing change notifications they should ALWAYS implement delta as the premium and reliable way to get the actual change data.

Author: not really, they could either rely on rich notifications when available or just read the full entity and overwrite their copy. The later might be more costly but it's simpler to implement (but I agree it doesn't guarantee data integrity, especially when you factor in partitions affinity and potential sync delays).

- API consumers are not looking for a "one-time" export or back-up mechanism.

### Alternatives

- Change notifications pattern (TODO add link when described)

> - Export pattern (TODO)

Reviewer: I'm not sure yet that backup/export  will end up as patterns.  Let's add them in if they end up existing.

Author: ```suggestion
```

Implementer MUST implement soft deletion for entities in the backend storage system. The soft deletion will provide useful information to the client to appropriately reflect deletions.

When an entity is soft deleted, the delta function MUST return the id of the deleted entity as well as a `@removed` annotation with the `reason` field.
- The reason MUST be set to `changed` if the entity can be restored. `"@removed": {"reason": "changed"}`. 
- The reason MUST be set to `deleted` if the entity cannot be restored. `"@removed": {"reason": "deleted"}`.

When a link to an entity is deleted, or when the linked entity is deleted, or when a link to an entity is added, implementer MUST return a `property@delta` annotation. e.g. considering the entity Group has a navigation property named members of type Collection(user):

> - When a user is added to the group `"members@delta": ["id of the added user"]`

Author: ```suggestion
- When a user is added to the group `"members@delta": [{ "@odata.type": "#microsoft.graph.user", "id of the added user"}]`
```

> - When a user is removed from the group, or the target user is deleted `"members@delta": ["id of the added user"]`

Author: ```suggestion
- When a user is removed from the group, or the target user is deleted `"members@delta": ["id of the added user"]`
```
```suggestion
- When a user is removed from the group, or the target user is deleted `"members@delta": [{"@removed": {"reason": "deleted"}, "id of the deleted or removed user"}]`
```

> # Change Tracking

Reviewer: I'd prefer we drop the term change tracking and call this delta from the get-go.  change tracking seems like an unnecessary odata-ism to use for the broad feature area.

Author: we call it change tracking in our public docs and training, I thought aligning on that was good in case they wanted to do more research and get more information.

Reviewer 2: I also think that Change Tracking is more generic and should stay  for consistency.

Author: great, resolving

Microsoft Graph API Design Pattern

> *The change tracking pattern provides the ability to keep a third party system in sync with changes in Microsoft Graph without having to continuously poll the API.*

Reviewer 3: ```suggestion
*The change tracking pattern provides the ability to keep API consumers in sync with changes in Microsoft Graph without having to continuously poll the API.*
```

## Problem
---------

> Third party systems require to keep data in sync with Microsoft Graph, and the API design should not allow for continuous polling as it'd be costly and wouldn't guarantee data integrity.

Reviewer 3: ```suggestion
API consumers require an efficient way to keep data in sync with Microsoft Graph, and the API design should not allow for continuous polling as it'd be costly and wouldn't guarantee data integrity.
```

> API consumers require an efficient way to keep data in sync with Microsoft Graph, and the API design should not allow for continuous polling as it'd be costly and wouldn't guarantee data integrity.

Reviewer: I think this needs a rewrite, because many customers do poll against the delta api pretty frequently.
It's not idea, but we don't have that many webhooks to make it unnecessary.

SO some language more like 'api design should be optimized to avoid polling as it is costly for consumers and producers alike.'

Author: ```suggestion
API consumers require an efficient way to keep data in sync with Microsoft Graph, and the API design should be optimized to avoid polling as it is costly for consumers and producers alike as well as wouldn't guarantee data integrity.
```

what do you think of this?

## Solution
--------

> API designers leverage the change tracking (delta) capability on the entity set and declare a delta function for API consumers to use when tracking changes happening in the system.

Reviewer 3: ```suggestion
API designers leverage the change tracking (delta) capability on the entity set and declares a delta function for API consumers to use when tracking changes happening in the system.
```

> API designers leverage the change tracking (delta) capability on the entity set and declare a delta function for API consumers to use when tracking changes happening in the system.

Reviewer 4: ```suggestion
API designers can enable the change tracking (delta) capability on entity collections by declaring a delta function for API consumers to use to track changes in that collection.
```

> This new endpoint can be used to sync API consumers. This is achieved through returning a delta link with a watermark. Once the API consumer needs to refresh the data it uses the last provided delta link to catch up on new changes since their last request. Delta guarantees integrity of data through the watermark, regardless of service partitions and other obscure aspects for clients.

Reviewer: Some explicit explanation needed around us choosing to standardize on use of a delta function rather than the odata standard to access the delta.  Doesn't need to justify why, rather explain that it is the case.

Also shoudl decide whether it is OK to support the odata standard in addition.

Author: ```suggestion
This new endpoint can be used to sync API consumers. This is achieved through returning a delta link with a watermark. Once the API consumer needs to refresh the data it uses the last provided delta link to catch up on new changes since their last request. Delta guarantees integrity of data through the watermark, regardless of service partitions and other obscure aspects for clients.

> Note: although this capability is similar to the [OData $delta feed](https://docs.oasis-open.org/odata/odata-json-format/v4.0/errata02/os/odata-json-format-v4.0-errata02-os-complete.html#_Toc403940644) capability, it is a different construct. Microsoft Graph APIs MUST provide change tracking through the delta function and MUST NOT implement the OData $delta feed when providing change tracking capabilities to ensure the uniformity of the API experience.
```

What do you think of that?

> This new endpoint can be uses to sync third party systems. This is achieved through returning a delta link with a watermark. Once the third party system comes back online, it uses the last provided delta link to catch up on new changes since their last request. Delta guarantees integrity of data through the watermark, regardless of service partitions and other obscure aspects for clients.

Reviewer 3: ```suggestion
This new endpoint can be used to sync API consumers This is achieved through returning a delta link with a watermark. When the API consumer needs to refresh the data, it uses the last provided delta link to catch up on new changes since their last request. Delta guarantees integrity of data through the watermark, regardless of service partitions and other obscure aspects for clients.
```

> This new endpoint can be used to sync third party systems. This is achieved through returning a delta link with a watermark. Once the third party system comes back online, it uses the last provided delta link to catch up on new changes since their last request. Delta guarantees integrity of data through the watermark, regardless of service partitions and other obscure aspects for clients.

Reviewer 3: ```suggestion
This new endpoint can be used to sync API consumers. This is achieved through returning a delta link with a watermark. Once the API consumer needs to refresh the data it uses the last provided delta link to catch up on new changes since their last request. Delta guarantees integrity of data through the watermark, regardless of service partitions and other obscure aspects for clients.
```

> ## Issues and Considerations

Reviewer 3: Should we mention that delta can be a composable function, therefore you can pass $filter etc.

Author: delta is not composable in the sense that it doesn't accept any path segment after (navigation properties, other functions, ...). But it does accept filter and select statements (I think that technically top and expand could be supported as well).

What do you think of that.

> Note: the delta function also provides support for $filter and $select to allow the API consumer to narrow down the number of entities and properties retrieved as well as the number of changes that are tracked. Additionally the delta function can also support $top to allow the API consumer to sync smaller sets of changes as well as $expand to allow the API consumer to sync related data. Expand across workloads is not supported today however.

-------------------------

> Implementer MUST implement a watermark storage system in case of active watermarks (cursor in data store, partition affinity, sync state in data store...).

Reviewer 3: Do we need to define what an ***active*** watermark is?

Author: ```suggestion
Implementer MUST implement a watermark storage system in case of active watermarks. Passive watermarks are watermarks that can be retrieved from the context (e.g. timestamp), active watermarks represent information required to track the sync state which cannot be retrieved from the context (e.g. cursor from data store, partition affinity marker, partition id, generated unique sync identifier...)
```

What do you think of that?

> 

Reviewer 3: Should also call out that the implementor should have a way to track when entities are deleted e.g. soft-delete patterns

Author: do you mean that so they are able to enumerate deletions when clients issue a request to the delta endpoint? usually the deleted entries only return their ids

Reviewer: I think you need to explain the structure of the @removed/reason part of the pattern here.

Author: added above

## When to Use this Pattern
------------------------

Before using the change tracking pattern in your API definition, make sure your scenario fits the following criteria:

- API consumers want to sync the data.

> - API consumers don't want to get notified of changes. (see change notifications pattern for this scenario).

Reviewer 3: ```suggestion
- API consumers don't want to be immediately notified of changes. (see change notifications pattern for this scenario).
```

### Alternatives

- Change notifications pattern (TODO add link when described)
- Export pattern (TODO)
- Backup pattern (TODO)

## Examples
-------

### Getting changes for the users entity set

```HTTP
GET https://graph.microsoft.com/v1.0/users/delta
```

```json
{
    "@odata.context": "https://graph.microsoft.com/v1.0/$metadata#users",
    "@odata.deltaLink": "https://graph.microsoft.com/v1.0/users/delta?$deltatoken=mS5DuRZGjVL-abreviated",
    "value": [
        {
            "businessPhones": [
                "+1 309 555 0104"
            ],
            "displayName": "Grady Archie",
            "givenName": "Grady",
            "jobTitle": "Designer",
            "mail": "GradyA@contoso.onmicrosoft.com",
            "officeLocation": "19/2109",
            "preferredLanguage": "en-US",
            "surname": "Archie",
            "userPrincipalName": "GradyA@contoso.onmicrosoft.com",
            "id": "0baaae0f-b0b3-4645-867d-742d8fb669a2",
            "manager@delta": [
                {
                    "@odata.type": "#microsoft.graph.user",
                    "id": "99789584-a1e1-4232-90e5-866170e3d4e7"
                }
            ]
        },

>     ]

Reviewer 3: We should have an example that covers something being added, something being updated, and something being removed

Author: added a removed example.
Added and updated are the same with the instructions to the client app to uppsert.

}
```

> > Note: the response contains an `@odata.nextLink` instance annotation with the watermark only when all the changes are enumerated. If more changes need to be enumerated, the response instead contains an `@odata.skipLink` instance annotation the application can request right away to get the next page.

Reviewer 3: Is ```@odata.nextLink``` right here, shouldn't it be ```@odata.deltaLink```

Author: ```suggestion
> Note: the response contains an `@odata.deltaLink` instance annotation with the watermark only when all the changes are enumerated. If more changes need to be enumerated, the response instead contains an `@odata.nextLink` instance annotation the application can request right away to get the next page.
```
